### ISSUE TITLE
Captcha on Login View.

### ISSUE THREAD

**Issue by Reve** [gogits/gogs#2932]
Hello,

I think there should be a Captcha for the login window since it can be prone to brute force attacks.

Thank you

**Comment by lefta** [comment]
I don't think so, captchas are a defense against bots, especially but not exclusive to account register bots. In the case of brute force attacks, it needs a lot of requests to get a chance to find anything. I think a limit / time / IP is enough to protect against it. Eventually, you may blacklist someone after an amount of failed login attempts.

I did not check if gogs is already protected, in any case it should be possible to do it from nginx/apache if using them as proxies.

**Comment by bkcsoft** [comment]
@Reve Name one site which has captcha on the login-view and I'll fix an optional login-captcha. Otherwise I'm agreeing with @lefta :smiley:

**Comment by Reve** [comment]
[http://imgur.com/bHlxa3W](url)

As you can see, Stash (or Bitbucket Server), offers this option. Also there are botnets that that can distribute such kind of attacks which will bypass your fail2ban solution.

**Comment by gronke** [comment]
Calculating the password hashes is expensive. Therefore login is a potential target for DoS. An IP based rate limit after the third attempt would be a good thing from that perspective.
